### PR TITLE
feat: support slurm REST API for issue #41

### PR DESCRIPTION
**Pull request by uddmorningsun**
slurmrestd needs unshare System V namespace permission in container. If not, slurmrestd will raise below following:

slurmrestd: fatal: Unable to unshare System V namespace: Operation not permitted

Signed-off-by: Chenyang Yan <memory.yancy@gmail.com>

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -112,10 +114,10 @@ ARG SLURM_TAG=slurm-21-08-6-1
 RUN set -ex \
     && git clone -b ${SLURM_TAG} --single-branch --depth=1 https://github.com/SchedMD/slurm.git \
     && pushd slurm \
-    && ./configure --prefix=/usr --sysconfdir=/etc/slurm \
+    && ./configure --prefix=/usr --sysconfdir=/etc/slurm --enable-slurmrestd \
         --with-mysql_config=/usr/bin --libdir=/usr/lib64 \
     && sed -e 's|#!/usr/bin/env python3|#!/usr/bin/python|' -i doc/html/shtml2html.py \
-    && make install \
+    && make -j $(nproc) install \
```
Comment (giovtorres) [inline Dockerfile:120]: Could we make this an ARG with a default?

File: tests/test_slurmrestapi.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from http import HTTPStatus
+from http.client import HTTPResponse
+
+import pytest
+from testinfra.backend.base import CommandResult
+
+
+def test_unauthorized(container_ip):
+    with pytest.raises(urllib.request.HTTPError, match="401: UNAUTHORIZED"):
+        urllib.request.urlopen("http://%s:6820/openapi" % container_ip)
```
Comment (giovtorres) [inline tests/test_slurmrestapi.py:12]: Could we use f-strings instead?

File: tests/test_slurmrestapi.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+
+@pytest.mark.parametrize("username", ["foouser"])
+def test_slurm_ping(host, container_ip, username):
+    token_output: CommandResult = host.run("scontrol token username=%s", username)
+    assert token_output.succeeded
+    jwt_token = token_output.stdout.strip().lstrip("SLURM_JWT=")
+
+    request_obj = urllib.request.Request(
+        "http://%s:6820/slurm/v0.0.37/ping" % container_ip,
```
Comment (giovtorres) [inline tests/test_slurmrestapi.py:22]: Same here for f-strings

File: tests/test_slurmrestapi.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        headers={"X-SLURM-USER-TOKEN": jwt_token, "X-SLURM-USER-NAME": username},
+    )
+    response: HTTPResponse = urllib.request.urlopen(request_obj)
+    assert response.status == HTTPStatus.OK and response.headers["Content-Type"].lower().startswith("application/json")
```
Comment (giovtorres) [inline tests/test_slurmrestapi.py:26]: Can this be split into two assertions?

**Review by giovtorres — COMMENTED** [review]
This looks really good! Thank you for this contribution! Just a couple minor feedback changes.

**Comment by uddmorningsun** [comment]
@giovtorres I have updated and rebased, and then I have added more use cases based on my usage and learning. Thanks review!

File: files/supervisord.conf [hunk 5]
```diff
@@ -88,3 +88,18 @@ stderr_logfile=/var/log/supervisor/slurmd.log
 stderr_logfile_maxbytes=1MB
 stderr_logfile_backups=5
 priority=100
+
+[program:slurmrestd]
+user=root
+command=/usr/sbin/slurmrestd -vvvvv 0.0.0.0:6820
```
Comment (giovtorres) [inline files/supervisord.conf:94]: Is this daemonized, or does it require the `-D` flag like the other slurm processes?

File: tests/test_slurmrestapi.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+
+@pytest.mark.parametrize("username", ["slurm", "root"])
+def test_slurm_ping(host, container_ip, username):
+    token_output: CommandResult = host.run("scontrol token username=%s", username)
+    assert token_output.succeeded
+    jwt_token = get_slurm_token(token_output.stdout)
+
+    request_obj = urllib.request.Request(
+        f"http://{container_ip}:6820/slurm/v0.0.37/ping",
```
Comment (giovtorres) [inline tests/test_slurmrestapi.py:28]: Is hardcoding the API version going to be a problem in the future? Is there a way we could fetch the API version used in the test in conftest.py, for example, similar to the container ip?

File: files/supervisord.conf (hunk 5 above)
Comment (uddmorningsun) [reply files/supervisord.conf:94]: Not, it's the foreground process, it will block

File: tests/test_slurmrestapi.py (hunk 6 above)
Comment (uddmorningsun) [reply tests/test_slurmrestapi.py:28]: Good idea, I will fetch and get latest API path to request. Details can refer to https://slurm.schedmd.com/rest_api.html#Openapi

**Review by giovtorres — APPROVED** [review]
LGTM 👍🏼  Thank you for addressing feedback!

**Pull request merged**